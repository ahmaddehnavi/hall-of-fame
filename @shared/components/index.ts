export * from './layout/Col'
export * from './layout/Row'
export * from './screen/Screen'
export * from './backhandler/MultiBackHandler'
export * from './touchable/Touchable'
export * from './button/Button'
export * from './list/FlatList'
export * from './empty/EmptyView'
export * from './loading/LoadingView'
export * from './error/ErrorView'
export * from './async-view/AsyncView'