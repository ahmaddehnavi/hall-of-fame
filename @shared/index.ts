export * from './services'
export * from './utils'
export * from './components'
export * from './resource/Resource'
