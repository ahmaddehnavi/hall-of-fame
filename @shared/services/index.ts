export * from './base'
export * from './navigation'
export * from './theme'

