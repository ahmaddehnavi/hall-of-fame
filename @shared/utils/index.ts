export * from './di/DIUtil'
export * from './async/AsyncUtil'