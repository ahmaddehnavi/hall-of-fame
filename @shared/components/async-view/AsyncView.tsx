import {Resource} from '@shared';
import {observer} from 'mobx-react';
import * as React from 'react';
import {StyleProp, UIManager, ViewProps, ViewStyle} from 'react-native';
import {View} from 'react-native-animatable';
import {ErrorView} from '../error/ErrorView';
import {LoadingView} from '../loading/LoadingView';

type ComponentWithProps<Req, DataType> = React.ComponentType<{ resource: Resource<Req, DataType>, style?: StyleProp<ViewStyle> }>
type AsyncViewProps<Req, DataType> = Partial<ViewProps> & {
    /**
     * if resource require 'request' info to load you should set it before pass to
     * this component or set `autoLoad` prop to 'never' and call load method yourself
     */
    autoLoad?: 'always' | 'no-data' | 'never'
    resource: Resource<Req, DataType>
    ErrorComponent?: ComponentWithProps<Req, DataType>
    LoadingComponent?: ComponentWithProps<Req, DataType>
    SuccessComponent?: ComponentWithProps<Req, DataType>
    children?: ComponentWithProps<Req, DataType> | ((res: Resource<Req, DataType>) => React.ReactNode)
};

@observer
export class AsyncView<Req, DataType> extends React.Component<AsyncViewProps<Req, DataType>> {

    static defaultProps = {
        ErrorComponent: ErrorView,
        LoadingComponent: LoadingView,
        autoLoad: 'no-data'
    };

    componentDidMount() {
        let autoLoad = this.props.autoLoad;
        let res = this.props.resource;
        // do nothing if loading already in progress
        if (res.isLoading) {
            return;
        }
        if (autoLoad === 'always') {
            this.props.resource.load()
        } else if (autoLoad === 'no-data' && !res.isSuccess) {
            this.props.resource.load()
        }
    }

    render() {
        let {
            style,
            ErrorComponent,
            LoadingComponent,
            SuccessComponent,
            resource,
            ...otherProps
        } = this.props;

        if (resource.isLoading && LoadingComponent) {
            return (
                <LoadingComponent
                    style={style}
                    resource={resource}/>
            )
        }
        if (resource.isError && ErrorComponent) {
            return (
                <ErrorComponent
                    style={style}
                    resource={resource}/>
            )
        }

        if (SuccessComponent) {
            return <SuccessComponent
                resource={resource}
                style={style}
                children={this.props.children}/>
        }
        return (
            this.props.children || false
        );
    }

}