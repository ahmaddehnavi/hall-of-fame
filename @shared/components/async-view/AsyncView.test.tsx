import {Resource} from '@shared';
import React from 'react';
import 'react-native';
import {Text} from 'react-native';
// Note: test renderer must be required after react-native.
import renderer from 'react-test-renderer';
import {AsyncView} from './AsyncView';

describe('AsyncView Tests', () => {

    it('should match snapshot', () => {
        let res = Resource.form(_ => Promise.resolve({data: 'test'}));
        let tree = renderer.create(
            <AsyncView
                resource={res}
            />
        );
        expect(tree).toMatchSnapshot();
    });


    it('should show data', async () => {
        let res = Resource.form(_ => Promise.resolve({data: 'test'}));
        await res.load();
        let tree = renderer.create(
            <AsyncView resource={res}>
                {
                    !!res.data &&
                    <Text>{res.data}</Text>
                }
            </AsyncView>
        );
        let text = tree.root.findByType(Text);
        expect(text.props.children).toBe('test');
    });
});