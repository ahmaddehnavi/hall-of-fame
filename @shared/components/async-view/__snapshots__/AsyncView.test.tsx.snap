// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AsyncView Tests should match snapshot 1`] = `
<View
  style={
    Array [
      Object {
        "flexDirection": "row",
      },
      Object {
        "alignItems": "center",
        "backgroundColor": "rgba(255,255,255,.5)",
        "flex": 1,
        "justifyContent": "center",
      },
    ]
  }
>
  <ActivityIndicator
    animating={true}
    color="#999999"
    hidesWhenStopped={true}
    size="small"
  />
  <Text
    style={
      Object {
        "marginStart": 8,
      }
    }
  >
    Loading...
  </Text>
</View>
`;
